evm_loader/program/src/token.rs

```diff
@@ -198,8 +207,162 @@ pub fn transfer_token(
+) -> Result<(), ProgramError> {
+
+    if gas_to_be_paid > gas_limit {
+        return Err!(ProgramError::InvalidArgument;
+                "gas_to_be_paid > gas_limit; gas_to_be_paid = {:?}; gas_limit = {:?}",
+                gas_to_be_paid, gas_limit);
+    }
+
+    if let Some(storage) = storage_opt {
+        let (gas_used_and_paid, number_of_payments) =
+            storage.get_payments_info()?;
+
+        msg!("gas_used_and_paid = {:?}; gas_to_be_paid={:?} by an iteration N = {:?}",
+            gas_used_and_paid, gas_to_be_paid, number_of_payments+1);
+
+        let gas_to_be_paid = if gas_to_be_paid < gas_used_and_paid
+        {
+            msg!("user does not pay for this iteration");
+            0
+        } else {
+            gas_to_be_paid.checked_sub(gas_used_and_paid)
+                .ok_or_else(|| E!(ProgramError::InvalidArgument))?
+        };
+
+        user_pays_operator_impl(
+            gas_price,
+            gas_to_be_paid,
+            user_token_account,
+            operator_token_account,
+            accounts,
+            account_storage,
+        )?;
+
+        let gas_has_been_paid = gas_to_be_paid;
+        debug_print!("user_pays_operator gas_has_been_paid = {:?}", gas_has_been_paid);
+        storage.add_gas_has_been_paid(gas_has_been_paid)
+    }
+    else {
+        user_pays_operator_impl(
+            gas_price,
+            gas_to_be_paid,
+            user_token_account,
+            operator_token_account,
+            accounts,
+            account_storage,
+        )
+    }
+}
+
+fn check_enough_funds_impl<'a>(
+    gas_limit: u64,
+    gas_price: u64,
+    gas_used_and_paid: u64,
+    user_token_account: &'a AccountInfo<'a>
+) -> Result<(bool, u64, u64, u64, u64), ProgramError> {
+
+    let user_balance_64 = get_token_account_balance(user_token_account)?;
+    let user_balance : U256 = U256::from(user_balance_64)
+        .checked_mul(eth::min_transfer_value())
+        .ok_or_else(|| E!(ProgramError::InvalidArgument))?;
+    let gas_price_wei = U256::from(gas_price);
+    let gas_to_be_paid = gas_limit.checked_sub(gas_used_and_paid)
+        .ok_or_else(|| E!(ProgramError::InvalidArgument))?;
+    let expected_fee = U256::from(gas_to_be_paid)
+        .checked_mul(gas_price_wei)
+        .ok_or_else(|| E!(ProgramError::InvalidArgument))?;
+    Ok((expected_fee < user_balance, gas_limit, gas_price, gas_used_and_paid, user_balance_64))
+}
+
+/// Check that neon-evm user has enough funds to pay for gas
+///
+/// # Errors
+///
+/// Could return:
+/// `ProgramError::InvalidArgument`
+pub fn check_enough_funds<'a>(
+    gas_limit: u64,
+    gas_price: u64,
+    user_token_account: &'a AccountInfo<'a>,
+    storage_opt: Option<&mut StorageAccount>,
+) -> Result<(), ProgramError> {
+
+    let gas_used_and_paid = if let Some(storage) = storage_opt {
+        storage.get_payments_info()?.1
+    }
+    else { 0 };
+
+    let enough_funds = check_enough_funds_impl(
```

Comment:
I see no need in this function.
Check `if ! enough_funds.0 { ...` should have been moved to `check_enough_funds_impl`. But in that case in this function only `gas_used_and_paid` acquiring remains.
So i think it is better to get rid of this function.